Add tslearn

rth opened the pull request:
This adds [tslearn](https://github.com/rtavenar/tslearn) -- a machine learning toolkit dedicated to time-series data 

keras is an optional dependency (needed for just one module). I decided not to add it by default, as it will pull tensorflow on Unix which is a pretty heavy dependency and only a fraction of users of this package need the corresponding functionality.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/tslearn```) and found some lint.

Here's what I've got...


For **recipes/tslearn**:

 * Failed to even lint the recipe (might be a conda-smithy bug) :cry:

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/tslearn```) and found it was in an excellent condition.

rth commented on recipes/tslearn/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+test:
+  imports:
+    - tslearn
+    - tslearn.metrics
```
This will import most of compiled cython modules

jakirkham replied:
Admittedly this is challenging, but is there any way we can run a dummy test experiment here just to make sure it is working ok? If not, I understand, but it would be good to try. :)

jakirkham commented on recipes/tslearn/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+  home: http://tslearn.readthedocs.io/
+  license: BSD-2-Clause
+  license_family: BSD
+  license_file: '{{ environ["RECIPE_DIR"] }}/LICENSE'
```
Any chance we can get this upstreamed?

rth replied:
Made a PR upstream https://github.com/rtavenar/tslearn/pull/25

jakirkham replied:
Thanks. Feel free to update in the feedstock on the next release whenever that happens.

jakirkham commented:
Looks good. Made a couple comments above.

rth commented:
Thanks for the review @jakirkham !

> Admittedly this is challenging, but is there any way we can run a dummy test experiment here just to make sure it is working ok? If not, I understand, but it would be good to try. :)

I added a pytest command to run a subset of the test suite that is fast enough (https://github.com/rtavenar/tslearn/issues/26). Currently I added a `command` to `tests` with `pytest [..options..] tslearn`
where `tslearn` is the source folder with tests, but the path must be wrong. Is there an environment variable similar to `RECIPE_DIR` but that points to the source dir? Or is there a better way of doing this (e.g. via a `run_tests.sh`) ?

I wasn't able to find a simple example of to run tests. [This](https://github.com/conda-forge/datreant.data-feedstock/blob/31135f9d026572567fe33beab11185ddd3ff04c2/recipe/run_test.py) doesn't sound very practical.  It might be worth adding a small section about the best practices when running tests in a feedstock...

rth commented:
`pytest  [...options...] --pyargs tslearn` could be the way to go, but it doesn't seem to work well with the `--ignore` flags https://github.com/pytest-dev/pytest/issues/3287

Unless some easy solution is found on the pytest side, I'm half tempted to just remove that command test line from this PR ...

rth commented:
After more thoughts, I removed the test experiment altogether.
 * to run a subset of the test suite with pytest https://github.com/pytest-dev/pytest/issues/3287 would need to be addressed, this is achievable but will happen on a larger timescale than this PR anyway
 * I'm doubtful as to the usefulness of copying e.g. some of the tests to `run_tests.py`, maintenance wise this is not great.

Besides there are quite a few Python+Cython packages on conda forge that only test imports and they seem to be doing OK.

jakirkham commented:
No worries @rth. Knew it was a big ask. Thanks for trying anyways. Maybe a note can be added to an issue on the feedstock once it's out.

jakirkham commented on recipes/tslearn/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+    - python
+    - numpy >=1.11
+    - scipy
+    - scikit-learn
```
Surprising that it is not more sensitive to scikit-learn version. Guess if there are issues, upstream will hear about them as well.

rth replied:
True. Well, also scikit-learn <0.17.1 is not currently available on conda forge, and 0.17.1 might be fine.

rth replied:
>  also scikit-learn <0.17.1 is not currently available on conda forge

Since it's possible to mix different conda channels, never mind my comment.

rth commented:
> Maybe a note can be added to an issue on the feedstock once it's out.

Will do. Thanks for the review @jakirkham !   (The CI is green)
